scripts/cnode-helper-scripts/cncli.sh

```diff
@@ -797,10 +809,187 @@ cncliPTsendslots() {
+  jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+  sed 's/"//g' >> "$tmpcsv"
+}
+
+runNextEpoch() {
+  getKoiosData
+  getCurrNextEpoch
+  echo "Processing next epoch: ${next_epoch}"
+  stake_param_next="--active-stake ${active_stake_mark} --pool-stake ${pool_stake_mark}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --ledger-set next ${stake_param_next} |
+  jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+  sed 's/"//g' >> "$tmpcsv"
+}
+
+runPreviousEpochs() {
+  [[ -z ${KOIOS_API} ]] && return 1
+
+  if ! pool_hist=$(curl -sSL -f "${KOIOS_API}/pool_history?_pool_bech32=${POOL_ID_BECH32}&_epoch_no=${epoch}" 2>&1); then
+    echo "ERROR: Koios pool_stake_snapshot history query failed."
+    return 1
+  fi
+
+  if ! epoch_hist=$(curl -sSL -f "${KOIOS_API}/epoch_info?_epoch_no=${epoch}" 2>&1); then
+    echo "ERROR: Koios epoch_stake_snapshot history query failed."
+    return 1
+  fi
+
+  pool_stake_hist=$(jq -r '.[].active_stake' <<< "${pool_hist}")
+  active_stake_hist=$(jq -r '.[].active_stake' <<< "${epoch_hist}")
+
+  echo "Processing previous epoch: ${epoch}"
+  stake_param_prev="--active-stake ${active_stake_hist} --pool-stake ${pool_stake_hist}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --epoch="${epoch}" --ledger-set prev ${stake_param_prev} |
+  jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+  sed 's/"//g' >> "$tmpcsv"
+
+  return 0
+}
+
+processAllEpochs() {
+  getCurrNextEpoch
+  IFS=' ' read -r -a epochs_array <<< "$EPOCHS"
+
+  for epoch in "${epochs_array[@]}"; do
+    if ! getConsensus "${epoch}"; then echo "ERROR: Failed to fetch protocol parameters for epoch ${epoch}."; return 1; fi
+    if [[ "$epoch" == "$curr_epoch" ]]; then
+      runCurrentEpoch
+    elif [[ "$epoch" == "$next_epoch" ]]; then
+      runNextEpoch
+    else
+      runPreviousEpochs
+    fi
+  done
+
+  id=1
+  while IFS= read -r row; do
+    echo "$id,$row" >> "$csvfile"
+    ((id++))
+  done < "$tmpcsv"
+
+  sqlite3 "$BLOCKLOG_DB" <<EOF
+DELETE FROM epochdata;
+VACUUM;
+.mode csv
+.import '$csvfile' epochdata
+REINDEX epochdata;
+EOF
+
+  row_count=$(sqlite3 "$BLOCKLOG_DB" "SELECT COUNT(*) FROM epochdata;")
+  echo "$row_count rows have been loaded into epochdata table in blocklog db"
+  echo "~ CNCLI epochdata table load completed ~"
+
+  rm $csvfile $tmpcsv
+}
+
+processSingleEpoch() {
+   getCurrNextEpoch
+   IFS=' ' read -r -a epochs_array <<< "$EPOCHS"
+
+   unset matched
+   for epoch in "${epochs_array[@]}"; do
+     [[ ${epoch} = "$1" ]] && matched=true && break
+   done
+   if [[ -z ${matched} ]]; then
+     echo -e "No slots found in blocklog table for epoch ${EPOCH}.\n"; return 1
```

Comment:
Should be $1 I think.

Reply:
I tried both "$1" and $1 and being the parameter only a single string of 3 digits number, I don't see the difference in behaviour, I can add the condition that epoch number must be 3 digits, still can be passed a larger number and get the message "No slots found in blocklog table for epoch "

I left it as $1 with no quotes anyway

Reply:
About

>waitFromEpochstart=21600 # if within 6 hours of new epoch don't run epochdata

maybe 6 hours are too much, but I noticed several times that pool_stake_snapshot point is intermittently available at least during the first three hours of a new epoch, altough I'm not sure about pool_history point

Reply:
> I tried both "$1" and $1 and being the parameter only a single string of 3 digits number, I don't see the difference in behaviour, I can add the condition that epoch number must be 3 digits, still can be passed a larger number and get the message "No slots found in blocklog table for epoch "
> 
> I left it as $1 with no quotes anyway

My comment here wasn't for the comparison but EPOCH variable in echo string.

Reply:
> About
> 
> > waitFromEpochstart=21600 # if within 6 hours of new epoch don't run epochdata
> 
> maybe 6 hours are too much, but I noticed several times that pool_stake_snapshot point is intermittently available at least during the first three hours of a new epoch, altough I'm not sure about pool_history point

I will check this and get back to you.

Reply:
i missed this, I'm going to check it

Reply:
Hello @Scitz0 

I just pushed the updates

>1. updated EPOCH for $1 in
echo -e "No slots found in blocklog table for epoch ${1}.\n"

>2. requoted $1 in condition  [[ ${epoch} = "$1" ]] && matched=true && break

Cheers
Manuel